// @accounts/graphql-api/lib/modules/accounts-password/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@graphql-modules/core");
var password_1 = require("@accounts/password");
var server_1 = require("@accounts/server");
var types_1 = tslib_1.__importDefault(require("./schema/types"));
var query_1 = tslib_1.__importDefault(require("./schema/query"));
var mutation_1 = tslib_1.__importDefault(require("./schema/mutation"));
var query_2 = require("./resolvers/query");
var mutation_2 = require("./resolvers/mutation");
var utils_1 = require("../../utils");
var core_2 = require("../core");
exports.AccountsPasswordModule = new core_1.GraphQLModule({
    name: 'accounts-password',
    typeDefs: function (_a) {
        var config = _a.config;
        return [types_1.default, query_1.default(config), mutation_1.default(config)];
    },
    resolvers: function (_a) {
        var _b;
        var config = _a.config;
        return (_b = {},
            _b[config.rootQueryName || 'Query'] = query_2.Query,
            _b[config.rootMutationName || 'Mutation'] = mutation_2.Mutation,
            _b);
    },
    imports: function (_a) {
        var config = _a.config;
        return [
            core_2.CoreAccountsModule.forRoot({
                userAsInterface: config.userAsInterface,
            }),
        ];
    },
    providers: function (_a) {
        var config = _a.config;
        return [
            {
                provide: server_1.AccountsServer,
                useValue: config.accountsServer,
            },
            {
                provide: password_1.AccountsPassword,
                useValue: config.accountsPassword,
            },
        ];
    },
    context: utils_1.context('accounts-password'),
    configRequired: true,
});
