// @accounts/server/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var accounts_server_1 = require("./accounts-server");
exports.AccountsServer = accounts_server_1.AccountsServer;
var tokens_1 = require("./utils/tokens");
exports.generateRandomToken = tokens_1.generateRandomToken;
var get_first_user_email_1 = require("./utils/get-first-user-email");
exports.getFirstUserEmail = get_first_user_email_1.getFirstUserEmail;
var server_hooks_1 = require("./utils/server-hooks");
exports.ServerHooks = server_hooks_1.ServerHooks;
var accounts_error_1 = require("./utils/accounts-error");
exports.AccountsJsError = accounts_error_1.AccountsJsError;
var errors_1 = require("./errors");
exports.AuthenticateWithServiceErrors = errors_1.AuthenticateWithServiceErrors;
exports.LoginWithServiceErrors = errors_1.LoginWithServiceErrors;
exports.ImpersonateErrors = errors_1.ImpersonateErrors;
exports.FindSessionByAccessTokenErrors = errors_1.FindSessionByAccessTokenErrors;
exports.RefreshTokensErrors = errors_1.RefreshTokensErrors;
exports.LogoutErrors = errors_1.LogoutErrors;
exports.ResumeSessionErrors = errors_1.ResumeSessionErrors;
exports.default = accounts_server_1.AccountsServer;
